src/imgui/imgui_raii_wrapper.h

```diff
@@ -0,0 +1,160 @@
+
+#include <imgui.h>
+
+namespace irw {
+
+class Windowzita {
+public:
+  [[nodiscard]] explicit Windowzita(std::string const& id) : selected{ImGui::Begin(id.c_str())} {}
+  Windowzita(Windowzita const&) = delete;
+  Windowzita& operator=(Windowzita const&) = delete;
+  Windowzita(Windowzita&&) = delete;
+  Windowzita& operator=(Windowzita&&) = delete;
+  ~Windowzita() {
+    ImGui::End();
+  }
+  explicit operator bool() & {
+    return selected;
+  }
+private:
+  bool selected{};
+};
+
+class TabBar {
+public:
+  [[nodiscard]] explicit TabBar(std::string const& id) : selected{ImGui::BeginTabBar(id.c_str())} {}
+  TabBar(TabBar const&) = delete;
+  TabBar& operator=(TabBar const&) = delete;
+  TabBar(TabBar&&) = delete;
+  TabBar& operator=(TabBar&&) = delete;
+  ~TabBar() {
+    ImGui::EndTabBar();
+  }
+  explicit operator bool() & {
+    return selected;
+  }
+private:
+  bool selected{};
+};
+
+class TabItem {
+public:
+  [[nodiscard]] explicit TabItem(std::string const& id) : selected{ImGui::BeginTabItem(id.c_str())} {}
+  TabItem(TabItem const&) = delete;
+  TabItem& operator=(TabItem const&) = delete;
+  TabItem(TabItem&&) = delete;
+  TabItem& operator=(TabItem&&) = delete;
+  ~TabItem() {
+    if (selected) {
+      ImGui::EndTabItem();
+    }
+  }
+  explicit operator bool() & {
+    return selected;
+  }
+private:
+  bool selected{};
+};
+
+class Text {
+public:
+  Text(char const* text_) : text{text_} {}
+  Text(std::string text_) : text{std::move(text_)} {}
+  Text(char const* text_, ImVec4 color_) : text{text_}, color{color_} {}
+  Text(std::string text_, ImVec4 color_) : text{std::move(text_)}, color{color_} {}
+  std::string get_text() const {
+    return text;
+  }
+  std::optional<ImVec4> get_color() const {
+    return color;
+  }
+  void write() const {
+    if (color) {
+      ImGui::TextColored(color.value(), "%s", text.c_str());
+    } else {
+      ImGui::Text("%s", text.c_str());
+    }
+  }
+private:
+  std::string text{};
+  std::optional<ImVec4> color{};
+};
+
+class Table {
+public:
+  using Row = std::vector<Text>;
+  class Builder {
+  public:
+    Builder& with_id(std::string name) {
+      id = std::move(name);
+      return *this;
+    }
+    Builder& with_number_of_columns(std::size_t n) {
+      number_of_columns = n;
+      return *this;
+    }
+    Builder& with_headers(Row row) {
+      headers = std::move(row);
+      return *this;
+    }
+    Builder& add_row(Row row) {
+      values.push_back(std::move(row));
+      return *this;
+    }
+    Table build() {
+      return Table{
+        id,
+        static_cast<int>(number_of_columns),
+        headers,
+        values
+      };
+    }
+  private:
+    std::string id{};
+    std::size_t number_of_columns{};
+    Row headers{};
+    std::vector<Row> values{};
+  };
+
+  explicit Table(std::string const& id, int number_of_columns, Row const& headers, std::vector<Row> const& values) : selected{ImGui::BeginTable(id.c_str(), number_of_columns)} {
```

Comment:
`    IMGUI_API void          EndTable();                                 // only call EndTable() if BeginTable() returns true!
`

I know you did not thought of error handling here, but I guess we have to throw if `selected` is not true while constructing `Table`, also because there's no point of constructing a `Table` if we don't are able to `BeginTable` to start with

Reply:
This comment only means that in the destructor we need "if (selected)". In imgui some begins always need end and some only need end if begin returns true. Gotta love these C-style APIs. I fixed this one and double checked all the others.

Reply:
I really think there's no strong reasoning to continue to build something when `ImGui::BeginWtvThing` returns `false`, if returning `false` means that the object wasn't properly constructed, we have to check really what they mean with this `false` because if that means that `ImGui` wasn't able to construct the GUI object properly, imho we should just stop the object construction there.

Reply:
We should have a logic that after building the obj and we only use it when `selected`is true. Also your ctor is toooo long. I'm on my way with the clang-formater check for pull request.

Reply:
then again guys, are we going to "build" an object on which the `ImGui::BeginWtvThing` returns `false`?
I don't see a reason to do that...please enlight me dear fellas.

Reply:
Ah, you're thinking that Begin returning false means that there was an error, but that's not it. For example, a BeginButton returns true if the button was pressed or false otherwise. For the BeginTab, it returns true if the tab is selected. Then we have the operator bool so we can do an if. Look at the file called test.cpp and you'll see. Give me a call if this is still confusing.

Reply:
ah ok, that was my reasoning.